File: GitToolBox/src/test/java/zielu/gittoolbox/cache/CacheTaskSchedulerTest.java
package zielu.gittoolbox.cache;

import static org.assertj.core.api.Assertions.assertThat;

import com.intellij.openapi.project.Project;
import git4idea.repo.GitRepository;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import zielu.gittoolbox.metrics.MockMetrics;

@Tag("fast")
@ExtendWith({MockitoExtension.class})
class CacheTaskSchedulerTest {
  @Mock(stubOnly = true)
  private GitRepository repository;
  @Mock(stubOnly = true)
  private Project project;

  private CacheTaskScheduler scheduler;

  @BeforeEach
  void before() {
    scheduler = new CacheTaskScheduler(project, new MockMetrics());
    scheduler.setTaskDelayMillis(30);
    scheduler.initialize();
    scheduler.opened();
  }

  @AfterEach
  void after() {
    scheduler.closed();
    scheduler.dispose();
  }

  @Test
  void scheduleOptional() throws InterruptedException {
    CountDownLatch executedCount = new CountDownLatch(1);
    CountDownLatch discardedCount = new CountDownLatch(1);
    scheduler.scheduleOptional(repository, new DummyTask(executedCount));
    scheduler.scheduleOptional(repository, new DummyTask(executedCount, discardedCount));

    assertThat(discardedCount.await(1, TimeUnit.SECONDS)).isTrue();
    assertThat(executedCount.await(1, TimeUnit.SECONDS)).isTrue();
  }

  @Test
  void scheduleMandatory() throws InterruptedException {
    CountDownLatch executedCount = new CountDownLatch(2);
    scheduler.scheduleMandatory(repository, new DummyTask(executedCount));
    scheduler.scheduleMandatory(repository, new DummyTask(executedCount));

    assertThat(executedCount.await(1, TimeUnit.SECONDS)).isTrue();
  }

  @Test
  void removeRepositoryIfNeverScheduled() {
    scheduler.removeRepository(repository);
  }

  private static final class DummyTask implements CacheTaskScheduler.Task {
    private final CountDownLatch runLatch;
    private final CountDownLatch discardedLatch;

    private DummyTask(CountDownLatch runLatch, CountDownLatch discardedLatch) {
      this.runLatch = runLatch;
      this.discardedLatch = discardedLatch;
    }

    private DummyTask(CountDownLatch runLatch) {
      this(runLatch, new CountDownLatch(0));
    }

    @Override
    public void run(@NotNull GitRepository repository) {
      runLatch.countDown();
    }

    @Override
    public void discarded(@NotNull GitRepository repository) {
      discardedLatch.countDown();
    }
  }
}

File: GitToolBox/src/test/kotlin/zielu/gittoolbox/cache/CacheTaskSchedulerTest.kt
package zielu.gittoolbox.cache

import git4idea.repo.GitRepository
import io.mockk.impl.annotations.MockK
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.junit5.MockKExtension
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import kotlin.test.assertTrue

@ExtendWith(MockKExtension::class)
internal class CacheTaskSchedulerTest {
  @RelaxedMockK
  private lateinit var gatewayMock: CacheTaskSchedulerLocalGateway
  private lateinit var scheduler: CacheTaskScheduler

  @BeforeEach
  fun beforeEach() {
    scheduler = CacheTaskScheduler(gatewayMock)
    scheduler.setTaskDelayMillis(30)
  }

  @AfterEach
  fun afterEach() {
    scheduler.dispose()
  }

  @Test
  @Throws(InterruptedException::class)
  fun `scheduleOptional should discard second task`(@MockK repository: GitRepository) {
    // given
    val executedCount = CountDownLatch(1)
    val discardedCount = CountDownLatch(1)

    // when
    scheduler.scheduleOptional(repository, DummyTask(executedCount))
    scheduler.scheduleOptional(repository, DummyTask(executedCount, discardedCount))

    // then
    assertSoftly { softly ->
      softly.assertThat(discardedCount.await(1, TimeUnit.SECONDS)).isTrue
      softly.assertThat(executedCount.await(1, TimeUnit.SECONDS)).isTrue
    }
  }

  @Test
  @Throws(InterruptedException::class)
  fun `scheduleMandatory should execute two tasks`(@MockK repository: GitRepository) {
    // given
    val executedCount = CountDownLatch(2)

    // when
    scheduler.scheduleMandatory(repository, DummyTask(executedCount))
    scheduler.scheduleMandatory(repository, DummyTask(executedCount))

    // then
    assertTrue { executedCount.await(1, TimeUnit.SECONDS) }
  }

  @Test
  fun `should remove repository if tasks were never scheduled`(@MockK repository: GitRepository) {
    scheduler.removeRepository(repository)
  }

  private class DummyTask(
    private val runLatch: CountDownLatch,
    private val discardedLatch: CountDownLatch = CountDownLatch(0)
  ) : CacheTaskScheduler.Task {
    override fun run(repository: GitRepository) = runLatch.countDown()
    override fun discarded() = discardedLatch.countDown()
  }
}
